ONSdigital/eq-questionnaire-runner
# Looping 4 - Calculated Summary with repeating blocks

### What is the context of this PR?
This PR adds the functionality that allows calculated summaries to use repeating block answers that resolve to a list of answers. 

Calculated summaries now support consisting of a single repeating answer, or both repeating and static answers where each repeating answer resolves to a list of answers for each list item. By repeating answer, this means either dynamic, or an answer for a repeating block.

This required a minor change to validator - [corresponding PR](https://github.com/ONSdigital/eq-questionnaire-validator/pull/206)

Note that when editing a list item from the section summary that is only used by the second calculated summary in the test schema - this will take you to the first calculated summary and then the second. Jumping to the second calculated summary will be covered by the card that takes you to the next incomplete block, rather than just stepping through the section until it is complete.

### Technical details
`placeholder_render.py` has been modified to suffix any repeating question blocks/answers with their list item id. This is just so the HTML outputted by the jinja templates is valid. To see this working, inspect element on the summary page of a list collector with repeating blocks and note that the div ids are the same for each list item id. With this change they should now have their list item id suffix (Just like dynamic answers would)

`questionnaire_schema.py` now stores a list of repeating block answer ids and a map of repeating blocks to their list, so that the value source resolver has the ability to check if an answer is for a repeating block, and if so, where to get the list of answers from

`value_source_resolver.py` now checks if an answer is either dynamic, or for a repeating block and for either case, calls a shared method to get the list of answers for each list item id.

`group.py` now checks if the block being rendered is a repeating question. It needs to do this before checking if the block is on the path, as list repeating questions won't be on it, its a property of the list collector instead. This will call a new method on `ListCollectorBlock` to return a list of blocks for the question, one copy for each list item id

`list_collector_block.py` Now accepts `return to` and `return_to_block_id` as this is now used outside of section summaries and also by calculated summaries. It has a new method to take a repeating question block and return a list of rendered copies for each list item id. It also renders related answers to ensure that any answer labels using placeholders show correctly.

`list_action.py`, `list_edit_question.py`, `list_repeating_question.py` have had various updates so that on a repeating block, the previous link will take you to the previous block or the edit block, and continuing will take you to the next incomplete repeating block, or back to a calculated/section summary if the block is complete and `return_to` is populated.

Unit tests have been added to cover the changes to value source resolver
integration tests have been added to cover the changes to routing on a list collector with repeating blocks
functional tests to cover any scenarios on the trello ticket that weren't covered by looping 4.5

### How to test functionally
Use the new `test_new_calculated_summary_repeating_blocks.json` schema to add list items with repeating blocks that then feed into two separate calculated summaries, one that also uses a static answer, and one that only has a single repeating answer. 
Check that only the correct answers show up and that routing works as expected when adding/removing/editing list items.

### How to review
Read through the trello card to familiarise with the concept and check that each scenario in the attached scenarios google sheet is working. Check the test coverage is suitable

### Checklist
